### PR TITLE
Locale configuration

### LINKED ISSUE #73: Ensure modules are translatable
Ensure string literals in custom templates and forms have `@translate` annotations or `translate()` functions around them. This will be tested with the extract tagged strings code that will extract from phtml, php and twig files.

### PR DESCRIPTION
**Pull request by stephenwf**
Closes #73 #80 #81 #82 #78

### CODE REVIEW

File: repos/shared-library/src/Utility/OmekaValue.php [hunk 1]
```diff
@@ -20,6 +20,12 @@ public static function translateValue($representation, $term, $lang = '')
         if (!$representation instanceof ItemRepresentation && !$representation instanceof ItemSetRepresentation) {
             throw new LogicException('Only Items and ItemSets can be translated.');
         }
+
+        // Examples of input/output:
+        // en-gb -> en
+        // en-us -> en
+        // en    -> en
+        // cy-gb -> cy
         $primaryLang = explode('-', $lang)[0] ?? $lang;
```
Comment (garyttierney) [inline repos/shared-library/src/Utility/OmekaValue.php]: We should use `Locale::parseLocale()` for this

Comment (stephenwf) [reply repos/shared-library/src/Utility/OmekaValue.php]: Changed that now to use both the parse and comparisons

File: repos/iiif-storage/src/Aggregate/DereferencedManifest.php [hunk 2]
```diff
@@ -70,11 +70,31 @@ public function mutate(ItemRequest $input)
...
+                    if ($summary) {
+                        $input->addFields(
+                            FieldValue::literalsFromRdf('dcterms:description', 'Summary', $summary)
+                        );
+                    }
+                }
+
+                $attribution = $input->getValue('sc:attributionLabel');
+                if (!$attribution || empty($attribution) || trim(current($attribution)->getValue()) === '') {
+                    $attribution = $manifest['attribution'] ?? null;
```
Comment (stephenwf) [inline repos/iiif-storage/src/Aggregate/DereferencedManifest.php]: Need some abstraction here like:
```
$input->replaceIfEmpty( ... );
```
to avoid this verbose block being used all over the place

**Comment by stephenwf** [comment]
Latest change:
Screenshot 2019-04-11 at 16 53 32

File: repos/iiif-storage/config/service.config.php [hunk 3]
```diff
@@ -403,11 +428,15 @@
...
             },
+            'html' => function (ContainerInterface $c) {
+                return $c->get(HtmlIngester::class);
+            },
         ],
     ],
     'media_renderers' => [
         'invokables' => [
             'iiif' => null,
+            'html' => null,
```
Comment (garyttierney) [inline repos/iiif-storage/config/service.config.php]: Why is this needed? I see we're replacing it as a factory, does an invokable take precedence?

File: repos/iiif-storage/src/Model/BuiltCanvas.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace IIIFStorage\Model;
+
+
+class BuiltCanvas
```
Comment (garyttierney) [inline repos/iiif-storage/src/Model/BuiltCanvas.php]: Would be nice to call this `Canvas`, since it's built by `CanvasBuilder`, but that leaves us with a naming conflict across another namespace.

File: repos/omeka-i18n-module/src/Mvc/Router/SkippableSegment.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    public function __construct($route, array $constraints = array(), array $defaults = array(), array $skippable = array())
+    {
+        $this->defaults = $defaults;
+        $this->skippable = $skippable;
+        $this->parts = $this->parseRouteDefinition($route);
+        $this->regex = $this->buildRegex($this->parts, $constraints);
+    }
+    /**
+     * factory(): defined by RouteInterface interface.
```
Comment (garyttierney) [inline repos/omeka-i18n-module/src/Mvc/Router/SkippableSegment.php]: Let's just do `@inheritdoc`

File: repos/omeka-i18n-module/src/Mvc/Router/SkippableSegment.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                case 'optional':
+                    $skippable = true;
+                    $optionalPart = $this->buildPath($part[1], $mergedParams, true, $hasChild, $options);
+                    if ($optionalPart !== '') {
+                        $path .= $optionalPart;
+                        $skip = false;
+                    }
+                    break;
+                case 'translated-literal':
+                    $path .= $translator->translate($part[1], $textDomain, $locale);
```
Comment (garyttierney) [inline repos/omeka-i18n-module/src/Mvc/Router/SkippableSegment.php]: Are we using this? Could be potentially useful.

File: repos/omeka-i18n-module/src/Site/LocalizationListener.php [hunk 7]
```diff
@@ -43,28 +45,44 @@ class LocalizationListener
...
 
         if (null === $locale || empty($locale)) {
             $locale = 'en';
         }
 
+        $isMultiLingual = $isOnSite ? boolval($this->siteSettings->get('i18n-multi-lingual-site')) : false;
+        /** @var TreeRouteStack $router */
+        $router = $event->getRouter();
+
+        if ($isMultiLingual && $routerLocale) {
```
Comment (garyttierney) [inline repos/omeka-i18n-module/src/Site/LocalizationListener.php]: Is setting it to null required if it's not present? If so we might as well capture this in a ternary: `$uriLocale = ($isMultilingual && $routerLocale) ? $locale : null;` 

Also, a nitpick: multilingual is one word, so shouldn't be uppercased at Lingual.

File: repos/shared-library/src/Utility/OmekaValue.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            if (self::langMatches($value->lang(), $lang)) {
+                $fallback = $value;
+            }
+        }
+
+        // Fallback found, then return that.
+        return $fallback;
+    }
+
+    static public function langMatches($langA, $langB)
```
Comment (garyttierney) [inline repos/shared-library/src/Utility/OmekaValue.php]: Inconsistent order of modifiers here, `public static` and `static public` in this function.

File: repos/shared-library/src/Utility/OmekaValue.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        if ($langA === $langB) {
+            return true;
+        }
+        $valueLanguage = Locale::parseLocale($langA)['language'] ?? null;
+
+        // Check if they match each other, in either direction.
+        return (
+            $langA &&
```
Comment (garyttierney) [inline repos/shared-library/src/Utility/OmekaValue.php]: Can we move this test for `$langA` up to an `if`  that exits early? Testing `$langA` isn't really part of the locale comparison and hurts the readability of this expression quite a bit.

File: repos/iiif-storage/src/JsonBuilder/MetadataAggregator.php [hunk 10]
```diff
@@ -2,54 +2,63 @@
...
-                        break;
-                    case 'literal':
-                        $json['metadata'][] = [
-                            'label' => $value->property()->label(),
-                            'value' => $value->value(),
-                        ];
-                        break;
-                    default:
-                        break;
+                // @todo Pass the labels through the translator in Omeka
```
Comment (garyttierney) [inline repos/iiif-storage/src/JsonBuilder/MetadataAggregator.php]: Hm, do we need to export vocabs to the external translation tool?

File: repos/iiif-storage/src/Model/BuiltCanvas.php (hunk 4 above)
Comment (stephenwf) [reply repos/iiif-storage/src/Model/BuiltCanvas.php]: Once we drop the IIIF-PHP library, we could do that. Its strange because there's also `CanvasRepresentation` which wraps IIIF-PHP, JSON and the Omeka item which is passed to the templates!

File: repos/omeka-i18n-module/src/Mvc/Router/SkippableSegment.php (hunk 6 above)
Comment (stephenwf) [reply repos/omeka-i18n-module/src/Mvc/Router/SkippableSegment.php]: No currently, hadn't noticed the extra type

File: repos/iiif-storage/src/JsonBuilder/MetadataAggregator.php (hunk 10 above)
Comment (stephenwf) [reply repos/iiif-storage/src/JsonBuilder/MetadataAggregator.php]: Omeka 1.4 added [something that might help](https://github.com/omeka/omeka-s/pull/1373) but this appears to be just for displaying the vocabulary. It's possible we might need to do this, but I don't think it is worth doing at the moment. Specifically, the labels I'm referring to in this `@todo` are the ones from the resource templates:
Screenshot 2019-04-11 at 19 40 52

Which is used in the metadata key-value pair on an omeka item. (bad example here showing `sc:hasManifests` but imagine another RDF term)

So maybe we just need to export the resource templates, and then pass them through translate in the template

File: repos/iiif-storage/config/service.config.php (hunk 3 above)
Comment (stephenwf) [reply repos/iiif-storage/config/service.config.php]: Possibly, I remember running into issues with the IIIF one, also to remove any ambiguity for anyone debugging. (Quick scan of Zend docs didn't produce an answer either)

File: repos/iiif-storage/src/JsonBuilder/MetadataAggregator.php (hunk 10 above)
Comment (garyttierney) [reply repos/iiif-storage/src/JsonBuilder/MetadataAggregator.php]: Ah yeah, that could be as simple as exporting message keys for `dcterms:identifier` or something.
